### PR TITLE
Feature/8103 add settings changes detection

### LINKED ISSUE #8103: [RI Recommendations] Task 2.3: Add Settings Change Detection
# Task 2.3: Add Settings Change Detection

**File:** `inc/Engine/Admin/RocketInsights/Recommendations/SettingsSubscriber.php`

## User Story
**As a** WP Rocket user  
**I want** fresh recommendations when I change relevant settings  
**So** suggestions reflect my current configuration

---

### Core Logic
- [ ] Detect when WP Rocket settings are saved
- [ ] Check if changed settings are in the recommendation options map
- [ ] Only trigger fetch if:
  - Current recommendation status is `completed` or `failed`
  - At least one changed setting affects recommendations
  - Metrics hash would change (data is different)

### Settings to Monitor
All settings in `DataManager::get_enabled_options()` map:
- Minification options
- JavaScript optimization
- CSS optimization
- Lazy loading options
- CDN settings
- Cache settings
- etc.

---

### Create New Subscriber

```php
<?php
declare(strict_types=1);

namespace WP_Rocket\Engine\Admin\RocketInsights\Recommendations;

use WP_Rocket\Event_Management\Subscriber_Interface;
use WP_Rocket\Logger\LoggerAware;
use WP_Rocket\Logger\LoggerAwareInterface;

/**
 * Subscriber for handling recommendation updates when settings change.
 *
 * @since 3.20.5
 */
class SettingsSubscriber implements Subscriber_Interface, LoggerAwareInterface {
    use LoggerAware;
    
    /**
     * DataManager instance.
     *
     * @var DataManager
     */
    private $data_manager;
    
    /**
     * Constructor.
     *
     * @param DataManager $data_manager DataManager instance.
     */
    public function __construct( DataManager $data_manager ) {
        $this->data_manager = $data_manager;
    }
    
    /**
     * {@inheritdoc}
     */
    public static function get_subscribed_events(): array {
        return [
            'rocket_after_save_options' => 'maybe_fetch_after_settings_change',
        ];
    }
    
    /**
     * Maybe fetch recommendations after settings change.
     *
     * Only fetches if:
     * - Status is completed or failed
     * - Changed settings affect recommendations
     * - Metrics hash would change
     *
     * @param array $old_options Previous settings.
     * @param array $new_options New settings.
     */
    public function maybe_fetch_after_settings_change( array $old_options, array $new_options ): void {
        // Check current status
        $status = $this->data_manager->get_status();
        
        // Only proceed if recommendations exist
        if ( ! in_array( $status, [ 'completed', 'failed' ], true ) ) {
            $this->logger::debug(
                'Recommendations: Settings changed but status not ready',
                [ 'status' => $status ]
            );
            return;
        }
        
        // Check if relevant settings changed
        if ( ! $this->has_relevant_changes( $old_options, $new_options ) ) {
            $this->logger::debug( 'Recommendations: Settings changed but none affect recommendations' );
            return;
        }
        
        // Check if hash would change (prevents redundant fetch)
        if ( ! $this->data_manager->should_fetch_recommendations() ) {
            $this->logger::debug( 'Recommendations: Settings changed but hash unchanged' );
            return;
        }
        
        $this->logger::info( 'Recommendations: Relevant settings changed, fetching new recommendations' );
        
        // Fetch new recommendations
        $this->data_manager->fetch_recommendations();
    }
    
    /**
     * Check if changed settings affect recommendations.
     *
     * @param array $old_options Previous settings.
     * @param array $new_options New settings.
     * @return bool True if relevant changes detected.
     */
    private function has_relevant_changes( array $old_options, array $new_options ): bool {
        // Get list of recommendation-related options
        $relevant_keys = [
            'minify_css',
            'minify_js',
            'minify_concatenate_css',
            'minify_concatenate_js',
            'defer_all_js',
            'delay_js',
            'async_css',
            'critical_css',
            'remove_unused_css',
            'lazyload',
            'lazyload_iframes',
            'lazyload_youtube',
            'image_dimensions',
            'cdn',
            'do_caching_mobile_files',
            'cache_logged_user',
            'cache_webp',
            'manual_preload',
            'sitemap_preload',
            'control_heartbeat',
            'minify_google_fonts',
        ];
        
        // Check if any relevant setting changed
        foreach ( $relevant_keys as $key ) {
            $old_value = $old_options[ $key ] ?? false;
            $new_value = $new_options[ $key ] ?? false;
            
            if ( $old_value !== $new_value ) {
                return true;
            }
        }
        
        return false;
    }
}
```

### Register in ServiceProvider
```
// Recommendations Settings Subscriber.
$this->getContainer()->addShared( 'recommendations_settings_subscriber', SettingsSubscriber::class )
    ->addArgument( 'recommendations_data_manager' );
```

### Register in Plugin.php
```
$subscribers = [
    // ...
    'recommendations_fetch_subscriber',
    'recommendations_settings_subscriber', // Add this
    // ...
];
```

### Unit Tests
- [ ] maybe_fetch_after_settings_change() with various status states
- [ ] has_relevant_changes() detects relevant changes
- [ ] has_relevant_changes() ignores irrelevant changes
- [ ] Hash check prevents redundant fetches
- [ ] Fetch only called when all conditions met

### Integration Tests
- [ ] Settings save triggers fetch when status is completed
- [ ] Settings save does nothing when status is pending
- [ ] Irrelevant setting changes don't trigger fetch
- [ ] Hash unchanged = no fetch even if setting changed

## Edge Cases
**Scenario 1:** User disables a feature that's in recommendations
**Result:** Fetch new recommendations (feature no longer relevant)

**Scenario 2:** User enables a feature that's in recommendations
**Result:** Fetch new recommendations (feature now active, shouldn't be recommended)

**Scenario 3:** User changes unrelated setting (e.g., cache lifespan)
**Result:** No fetch (not in recommendations scope)

**Scenario 4:** Settings changed while status is pending
**Result:** No fetch (wait for tests to complete first)

### PR DESCRIPTION
**Pull request by Miraeld**
# Description

Fixes #8103 
This pull request introduces a new mechanism to automatically refresh performance recommendations when relevant WP Rocket settings change. It ensures that recommendations are only fetched when necessary, based on changes to key options or metrics, and adds robust unit and integration tests for this logic. The most important changes are as follows:

## Type of change

- [x] New feature (non-breaking change which adds functionality).

### What was tested

**Automated Tests (✅ Passing):**
- Unit tests for `SettingsSubscriber::maybe_fetch_after_settings_change()` - 8 scenarios covering all status conditions (pending, loading, completed, failed), relevant vs irrelevant changes, and hash-based optimization
- Unit tests for `SettingsSubscriber::get_subscribed_events()` - Verifies hook registration with correct parameters
- Unit tests for `DataManager::has_required_metrics()` - 6 scenarios covering all core metrics (LCP, TTFB, CLS, TBT) validation
- Integration tests for settings change detection with WordPress hooks and transient operations

**Manual Tests (Recommended):**
1. **Relevant Setting Change Triggers Fetch** - Toggle "Minify CSS" → Save → Verify new recommendations fetched
2. **Irrelevant Setting Change Does NOT Trigger** - Change "Cache Lifespan" → Save → Verify no fetch occurs
3. **No Fetch When Status Pending** - Clear recommendations transient → Change setting → No fetch until status is completed
4. **Multiple Settings Changed** - Toggle multiple options at once → Single fetch triggered
5. **Hash Unchanged Prevention** - Save same settings twice → Second save doesn't trigger redundant fetch

### How to test

**Prerequisites:**
1. Enable Rocket Insights with active license
2. Ensure Global Score is completed (status = "completed") - Visit Dashboard → Rocket Insights, test at least one URL
3. Optional: Enable `WP_DEBUG_LOG` to monitor log entries in `/wp-content/debug.log`

**Test 1: Relevant Setting Change Triggers Fetch**

Steps:
1. Go to **Settings → WP Rocket → File Optimization**
2. Toggle **Minify CSS files** (OFF → ON or ON → OFF)
3. Click **Save Changes**

Expected: 
- Settings save successfully
- New recommendations are fetched from API
- Transient `wpr_ri_recommendations` updated with new `metrics_hash`
- Log shows: `"Recommendations: Relevant settings changed, fetching new recommendations"`

**Test 2: Irrelevant Setting Change Does NOT Trigger Fetch**

Steps:
1. Go to **Settings → WP Rocket → Advanced Rules**
2. Change **Cache Lifespan** (e.g., 10h → 20h)
3. Click **Save Changes**

Expected:
- Settings save successfully
- NO fetch occurs
- Log shows: `"Recommendations: Settings changed but none affect recommendations"`

**Test 3: Multiple Relevant Settings at Once**

Steps:
1. Go to **Settings → WP Rocket → File Optimization**
2. Change multiple settings: Enable **Minify CSS**, **Minify JavaScript**, **Combine JavaScript**
3. Click **Save Changes**

Expected:
- Single fetch triggered (not multiple)
- New recommendations reflect all changes

**Monitored Settings (21 total):**
- Minify CSS/JS, Combine CSS/JS, Defer JS, Delay JS, Async CSS, Critical CSS, Remove Unused CSS
- LazyLoad (Images, Iframes, YouTube), Image Dimensions
- CDN, Mobile Cache, Cache Logged Users, WebP Cache
- Manual Preload, Sitemap Preload, Heartbeat Control, Google Fonts

**Verification Methods:**
```bash
# Check transient via WP-CLI
wp transient get wpr_ri_recommendations

# Check for metrics_hash field
wp transient get wpr_ri_recommendations | jq .metrics_hash

# Monitor logs (if WP_DEBUG_LOG enabled)
tail -f wp-content/debug.log | grep "Recommendations:"
```

### Affected Features & Quality Assurance Scope

**Primary Feature:**
- Rocket Insights → Recommendations System
- Settings change detection and automatic recommendations refresh

### Documentation

**Feature: Automatic Recommendations Refresh on Settings Change**
- Added a new `SettingsSubscriber` class that listens for changes to WP Rocket settings and triggers a recommendations refresh only if relevant settings are changed and the underlying metrics hash differs, preventing unnecessary API calls.
- Registered the new subscriber in the service provider and plugin initialization to ensure it is loaded and active in the application lifecycle. [[1]](diffhunk://#diff-304c3d096ea5ce72d2acc923ce3f7f69fd91c364b7b9853aedc4cb1033b2e74fR20) [[2]](diffhunk://#diff-304c3d096ea5ce72d2acc923ce3f7f69fd91c364b7b9853aedc4cb1033b2e74fR64) [[3]](diffhunk://#diff-304c3d096ea5ce72d2acc923ce3f7f69fd91c364b7b9853aedc4cb1033b2e74fR211-R214) [[4]](diffhunk://#diff-b591480fdcd92670a10d4e86f4f973f95ad761f81d44ce8f04109579a10dfebfR417)

**Enhancement: Metrics and Hash Validation**
- Added methods to `DataManager` for checking if required metrics exist and for determining if recommendations should be fetched, based on a hash of current metrics and enabled options. This helps avoid redundant recommendation fetches.
- When saving recommendations, now includes the metrics hash in the cached data for accurate comparison on future settings changes. (Fb9cb8f5L261R345)

**Testing: Unit, Fixture, and Integration Coverage**
- Added fixture and unit tests for `has_required_metrics` in `DataManager`, covering all edge cases of metric presence. [[1]](diffhunk://#diff-622055caff1760ece11617cfe748c58605a430b0356775cda435e817a9e311e6R1-R109) [[2]](diffhunk://#diff-f8de6f412b721443703e15c915f4d4e6adc33c80d807539034190118d9ca34d1R1-R48)
- Added fixture, unit, and integration tests for the new subscriber’s logic, ensuring recommendations are only fetched when appropriate and that the event subscription is correct. [[1]](diffhunk://#diff-0326bb9ff6f27d4d90f0ad254dc3674e6ea82affdc12bb0016d8595f94717cedR1-R224) [[2]](diffhunk://#diff-e92b9fdfa4722fb8cd118d9a6c346e362b7e16a19f91764a22d322aed3866fa4R1-R83) [[3]](diffhunk://#diff-ca22d1af81a77f6ce52930b324d4c1b1cc35b69f220314d10d341cec86ca4d58R1-R22)

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

### CODE REVIEW

File: inc/admin/options.php [hunk 1]
```diff
@@ -20,6 +20,16 @@ function rocket_after_save_options( $oldvalue, $value ) {
...
 
+	/**
+	 * Fires after WP Rocket settings are saved, before any cache operations
+	 *
+	 * @since 3.20.5
+	 *
+	 * @param array $oldvalue An array of previous values for the settings.
+	 * @param array $value    An array of submitted values for the settings.
+	 */
+	do_action( 'rocket_after_save_options', $oldvalue, $value );
```
Comment (wordpressfan) [inline inc/admin/options.php]: Why do we need this new action?
normally we use the action `update_option_wp_rocket_settings` exactly as we use it here:

https://github.com/wp-media/wp-rocket/blob/2fed94c5937eac74a1da1c8a5c9af6326003e844/inc/Engine/Tracking/Subscriber.php#L33

File: inc/Engine/Admin/RocketInsights/Recommendations/DataManager.php [hunk 2]
```diff
@@ -236,6 +236,90 @@ public function get_status(): string {
 		return $data['status'];
 	}
 
+	/**
```
Comment (wordpressfan) [inline inc/Engine/Admin/RocketInsights/Recommendations/DataManager.php:301]: this file changes are already there in [task 2.1](https://github.com/wp-media/wp-rocket/pull/8131)
This will cause conflicts, if it's needed we can merge Task 2.1 before this one.

File: inc/Engine/Admin/RocketInsights/Recommendations/SettingsSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	/**
+	 * Check if changed settings affect recommendations.
+	 *
+	 * @param array $old_options Previous settings.
+	 * @param array $new_options New settings.
+	 * @return bool True if relevant changes detected.
+	 */
+	private function has_relevant_changes( array $old_options, array $new_options ): bool {
+		// Get list of recommendation-related options.
+		$relevant_keys = [
```
Comment (wordpressfan) [inline inc/Engine/Admin/RocketInsights/Recommendations/SettingsSubscriber.php]: Can we unify the place that we get the options from, Now we have a method in DataManager::get_enabled_options to get enabled options, we may have a constant there and a method to get the list of tracked options and use it in both places, what do u think?

**Review by wordpressfan — CHANGES_REQUESTED** [review]

File: inc/Engine/Admin/RocketInsights/Recommendations/DataManager.php (hunk 2 above)
Comment (Miraeld) [reply inc/Engine/Admin/RocketInsights/Recommendations/DataManager.php:301]: I've removed these functions

**Review by wordpressfan — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@Miraeld Thanks for the PR.
[WIP] Exploratory test notes
- [ ] When changing settings, i.e., enable DJS, we delete the transient `wpr_ri_recommendations` and recreate it with the same values, including DJS, and only the timestamp is changed => https://group-onecom.slack.com/archives/C0AF91DUS9Z/p1773394256424069

**Comment by wordpressfan** [comment]
Fixed @Mai-Saad 
Going to merge this one to see everything in one place as we discussed today.
